In [8]:
import pandas as pd 
import funciones
import numpy as np

#### Extraccion y lectura de los datos

In [2]:
df_accidentes = pd.read_csv("../data/AccidentesAviones.csv", sep=",", encoding="utf-8")

Primera exploracion y estudi de la estrucrura del dataframe 

In [3]:
df_accidentes.head()

Unnamed: 0               fecha HORA declarada  \
0           0  September 17, 1908           1718   
1           1  September 07, 1909              ?   
2           2       July 12, 1912           0630   
3           3     August 06, 1913              ?   
4           4  September 09, 1913           1830   

                                 Ruta                OperadOR flight_no  \
0                 Fort Myer, Virginia    Military - U.S. Army         ?   
1             Juvisy-sur-Orge, France                       ?         ?   
2           Atlantic City, New Jersey    Military - U.S. Navy         ?   
3  Victoria, British Columbia, Canada                 Private         ?   
4                  Over the North Sea  Military - German Navy         ?   

           route                 ac_type registration cn_ln all_aboard  \
0  Demonstration        Wright Flyer III            ?     1          2   
1       Air show          Wright Byplane          SC1     ?          1   
2    Test flight               Dirigible            ?     ?          5   
3              ?        Curtiss seaplane            ?     ?          1   
4              ?  Zeppelin L-1 (airship)            ?     ?         20   

  PASAJEROS A BORDO crew_aboard cantidad de fallecidos passenger_fatalities  \
0                 1           1                      1                    1   
1                 0           1                      1                    0   
2                 0           5                      5                    0   
3                 0           1                      1                    0   
4                 ?           ?                     14                    ?   

  crew_fatalities ground                                            summary  
0               0      0  During a demonstration flight, a U.S. Army fly...  
1               0      0  Eugene Lefebvre was the first pilot to ever be...  
2               5      0  First U.S. dirigible Akron exploded just offsh...  
3               1      0  The first fatal airplane accident in Canada oc...  
4               ?      0  The airship flew into a thunderstorm and encou...

In [4]:
df_accidentes.shape

(5008, 18)

In [5]:
df_accidentes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5008 entries, 0 to 5007
Data columns (total 18 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   Unnamed: 0              5008 non-null   int64 
 1   fecha                   5008 non-null   object
 2   HORA declarada          5008 non-null   object
 3   Ruta                    5008 non-null   object
 4   OperadOR                5008 non-null   object
 5   flight_no               5008 non-null   object
 6   route                   5007 non-null   object
 7   ac_type                 5008 non-null   object
 8   registration            5008 non-null   object
 9   cn_ln                   5008 non-null   object
 10  all_aboard              5008 non-null   object
 11  PASAJEROS A BORDO       5008 non-null   object
 12  crew_aboard             5008 non-null   object
 13  cantidad de fallecidos  5008 non-null   object
 14  passenger_fatalities    5008 non-null   object
 15  crew

Por ahora nos encontramos con que el dataframe tiene 5008 registros, y 18 columnas, a a primera vista nos damos cuenta que tiene datos faltantes o nulos pero que estan representados con el caracter "?", por ahora evaluaremos el tipo de datos y el porcentaje de datos nulos antes de comenzar a reemplazar y transformar

In [6]:
funciones.verificar_datos(df_accidentes)

Advertencia: la columna route, tiene valores nulos


columna                         tipo_dato  datos_nulos  \
0               Unnamed: 0                   [<class 'int'>]            0   
1                    fecha                   [<class 'str'>]            0   
2           HORA declarada                   [<class 'str'>]            0   
3                     Ruta                   [<class 'str'>]            0   
4                 OperadOR                   [<class 'str'>]            0   
5                flight_no                   [<class 'str'>]            0   
6                    route  [<class 'str'>, <class 'float'>]            1   
7                  ac_type                   [<class 'str'>]            0   
8             registration                   [<class 'str'>]            0   
9                    cn_ln                   [<class 'str'>]            0   
10              all_aboard                   [<class 'str'>]            0   
11       PASAJEROS A BORDO                   [<class 'str'>]            0   
12             crew_aboard                   [<class 'str'>]            0   
13  cantidad de fallecidos                   [<class 'str'>]            0   
14    passenger_fatalities                   [<class 'str'>]            0   
15         crew_fatalities                   [<class 'str'>]            0   
16                  ground                   [<class 'str'>]            0   
17                 summary                   [<class 'str'>]            0   

    porcentaje_nulos  porcentaje_no_nulos  
0               0.00               100.00  
1               0.00               100.00  
2               0.00               100.00  
3               0.00               100.00  
4               0.00               100.00  
5               0.00               100.00  
6               0.02                99.98  
7               0.00               100.00  
8               0.00               100.00  
9               0.00               100.00  
10              0.00               100.00  
11              0.00               100.00  
12              0.00               100.00  
13              0.00               100.00  
14              0.00               100.00  
15              0.00               100.00  
16              0.00               100.00  
17              0.00               100.00

Podemos observear que eviste un solo dato nulo en la columna "rute", procedemos a verificar el dato

In [7]:
funciones.valores_nulos_columna(df_accidentes, "route")

c:\Users\fredd\Desktop\Data_Analytics_Accidentes_Aereos\Notebooks\funciones.py:38: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_nulos["index"] = df_nulos.index


Unnamed: 0           fecha HORA declarada               Ruta OperadOR  \
3173        3173  March 19, 1982          10:30  Leesburg, Florida  Private   

     flight_no route                 ac_type registration   cn_ln all_aboard  \
3173         ?   NaN  Beechcraft Bonanza F35       N567LT  D-4114          3   

     PASAJEROS A BORDO crew_aboard cantidad de fallecidos  \
3173                 2           1                      3   

     passenger_fatalities crew_fatalities ground  \
3173                    2               1      0   

                                                summary  index  
3173  Randy Rhoads, 25, lead guitarist for heavy met...   3173

Correcto el dato que figura como nulo en la columna "route" no es un registro totalmente vacio. 

Ahora procredemos a reemplazar el valor "?" por NaN y asi obtener com presición la cantidad de datos nulos del conjunto

In [9]:
df_accidentes.replace("?", np.nan, inplace= True)

In [10]:
df_accidentes.head()

Unnamed: 0               fecha HORA declarada  \
0           0  September 17, 1908           1718   
1           1  September 07, 1909            NaN   
2           2       July 12, 1912           0630   
3           3     August 06, 1913            NaN   
4           4  September 09, 1913           1830   

                                 Ruta                OperadOR flight_no  \
0                 Fort Myer, Virginia    Military - U.S. Army       NaN   
1             Juvisy-sur-Orge, France                     NaN       NaN   
2           Atlantic City, New Jersey    Military - U.S. Navy       NaN   
3  Victoria, British Columbia, Canada                 Private       NaN   
4                  Over the North Sea  Military - German Navy       NaN   

           route                 ac_type registration cn_ln all_aboard  \
0  Demonstration        Wright Flyer III          NaN     1          2   
1       Air show          Wright Byplane          SC1   NaN          1   
2    Test flight               Dirigible          NaN   NaN          5   
3            NaN        Curtiss seaplane          NaN   NaN          1   
4            NaN  Zeppelin L-1 (airship)          NaN   NaN         20   

  PASAJEROS A BORDO crew_aboard cantidad de fallecidos passenger_fatalities  \
0                 1           1                      1                    1   
1                 0           1                      1                    0   
2                 0           5                      5                    0   
3                 0           1                      1                    0   
4               NaN         NaN                     14                  NaN   

  crew_fatalities ground                                            summary  
0               0      0  During a demonstration flight, a U.S. Army fly...  
1               0      0  Eugene Lefebvre was the first pilot to ever be...  
2               5      0  First U.S. dirigible Akron exploded just offsh...  
3               1      0  The first fatal airplane accident in Canada oc...  
4             NaN      0  The airship flew into a thunderstorm and encou...

Podemos observar que todos los valores que estaban representados con "?" ahora estan como valores nulos y pueden ser contavilizados

In [11]:
funciones.verificar_datos(df_accidentes)

Advertencia: la columna HORA declarada, tiene valores nulos
Advertencia: la columna Ruta, tiene valores nulos
Advertencia: la columna OperadOR, tiene valores nulos
Advertencia: la columna flight_no, tiene valores nulos
Advertencia: la columna route, tiene valores nulos
Advertencia: la columna ac_type, tiene valores nulos
Advertencia: la columna registration, tiene valores nulos
Advertencia: la columna cn_ln, tiene valores nulos
Advertencia: la columna all_aboard, tiene valores nulos
Advertencia: la columna PASAJEROS A BORDO, tiene valores nulos
Advertencia: la columna crew_aboard, tiene valores nulos
Advertencia: la columna cantidad de fallecidos, tiene valores nulos
Advertencia: la columna passenger_fatalities, tiene valores nulos
Advertencia: la columna crew_fatalities, tiene valores nulos
Advertencia: la columna ground, tiene valores nulos
Advertencia: la columna summary, tiene valores nulos


columna                         tipo_dato  datos_nulos  \
0               Unnamed: 0                   [<class 'int'>]            0   
1                    fecha                   [<class 'str'>]            0   
2           HORA declarada  [<class 'str'>, <class 'float'>]         1504   
3                     Ruta  [<class 'str'>, <class 'float'>]            5   
4                 OperadOR  [<class 'str'>, <class 'float'>]           10   
5                flight_no  [<class 'float'>, <class 'str'>]         3682   
6                    route  [<class 'str'>, <class 'float'>]          763   
7                  ac_type  [<class 'str'>, <class 'float'>]           13   
8             registration  [<class 'float'>, <class 'str'>]          272   
9                    cn_ln  [<class 'str'>, <class 'float'>]          667   
10              all_aboard  [<class 'str'>, <class 'float'>]           17   
11       PASAJEROS A BORDO  [<class 'str'>, <class 'float'>]          221   
12             crew_aboard  [<class 'str'>, <class 'float'>]          219   
13  cantidad de fallecidos  [<class 'str'>, <class 'float'>]            8   
14    passenger_fatalities  [<class 'str'>, <class 'float'>]          235   
15         crew_fatalities  [<class 'str'>, <class 'float'>]          235   
16                  ground  [<class 'str'>, <class 'float'>]           44   
17                 summary  [<class 'str'>, <class 'float'>]           59   

    porcentaje_nulos  porcentaje_no_nulos  
0               0.00               100.00  
1               0.00               100.00  
2              30.03                69.97  
3               0.10                99.90  
4               0.20                99.80  
5              73.52                26.48  
6              15.24                84.76  
7               0.26                99.74  
8               5.43                94.57  
9              13.32                86.68  
10              0.34                99.66  
11              4.41                95.59  
12              4.37                95.63  
13              0.16                99.84  
14              4.69                95.31  
15              4.69                95.31  
16              0.88                99.12  
17              1.18                98.82